seed/seed16_test.go

```diff
@@ -0,0 +1,1058 @@
+			Required: true,
+			DevMode:  true,
+		},
+	})
+}
+
+func (s *seed16Suite) TestLoadMetaCore18StoreInfo(c *C) {
+	s.makeSeed(c, map[string]interface{}{
+		"base":   "core18",
+		"kernel": "pc-kernel=18",
+		"gadget": "pc=18",
+	}, snapdSeed, core18Seed, kernel18Seed, gadget18Seed, privateSnapSeed, contactableSnapSeed)
+
+	err := s.seed16.LoadAssertions(s.db, s.commitTo)
+	c.Assert(err, IsNil)
+
+	err = s.seed16.LoadMeta(s.perfTimings)
+	c.Assert(err, IsNil)
+
+	essSnaps := s.seed16.EssentialSnaps()
+	c.Check(essSnaps, HasLen, 4)
+
+	runSnaps, err := s.seed16.ModeSnaps("run")
+	c.Assert(err, IsNil)
+	c.Check(runSnaps, HasLen, 2)
+
+	privateSnapSideInfo := s.AssertedSnapInfo("private-snap").SideInfo
+	privateSnapSideInfo.Private = true
+	contactableSnapSideInfo := s.AssertedSnapInfo("contactable-snap").SideInfo
+	contactableSnapSideInfo.Contact = "author@example.com"
+
+	// these are not sorted by type, firstboot will do that
+	c.Check(runSnaps, DeepEquals, []*seed.Snap{
+		{
+			Path:     s.expectedPath("private-snap"),
+			SideInfo: &privateSnapSideInfo,
+			Channel:  "stable",
+		}, {
+			Path:     s.expectedPath("contactable-snap"),
+			SideInfo: &contactableSnapSideInfo,
+			Channel:  "stable",
+		},
+	})
+}
+
+func (s *seed16Suite) TestLoadMetaBrokenSeed(c *C) {
+	seedSnap16s := s.makeSeed(c, map[string]interface{}{
+		"base":           "core18",
+		"kernel":         "pc-kernel=18",
+		"gadget":         "pc=18",
+		"required-snaps": []interface{}{"required18"},
+	}, snapdSeed, core18Seed, kernel18Seed, gadget18Seed, required18Seed)
+
+	otherSnapFile := snaptest.MakeTestSnapWithFiles(c, `name: other
+version: other`, nil)
+	otherFname := filepath.Base(otherSnapFile)
+	err := os.Rename(otherSnapFile, filepath.Join(s.seedDir, "snaps", otherFname))
+	c.Assert(err, IsNil)
+
+	const otherBaseGadget = `name: pc
+type: gadget
+base: other-base
+version: other-base
+`
+	otherBaseGadgetFname, obgDecl, obgRev := s.MakeAssertedSnap(c, otherBaseGadget, nil, snap.R(3), "canonical")
+	s.WriteAssertions("other-gadget.asserts", obgDecl, obgRev)
+
+	err = s.seed16.LoadAssertions(s.db, s.commitTo)
+	c.Assert(err, IsNil)
+
+	omit := func(which int) func([]*seed.Snap16) []*seed.Snap16 {
+		return func(snaps []*seed.Snap16) []*seed.Snap16 {
+			broken := make([]*seed.Snap16, 0, len(snaps)-1)
+			for i, sn := range snaps {
+				if i == which {
+					continue
+				}
+				broken = append(broken, sn)
+			}
+			return broken
+		}
+	}
+	replaceFile := func(snapName, fname string) func([]*seed.Snap16) []*seed.Snap16 {
+		return func(snaps []*seed.Snap16) []*seed.Snap16 {
+			for i := range snaps {
+				if snaps[i].Name != snapName {
+					continue
+				}
+				sn := *snaps[i]
+				sn.File = fname
+				snaps[i] = &sn
+			}
+			return snaps
+		}
+	}
+
+	tests := []struct {
+		breakSeed func([]*seed.Snap16) []*seed.Snap16
+		err       string
+	}{
+		{omit(0), `model requires but seed is missing essential snap "snapd"`},
```

Comment:
I was initially wondering if using the name here instead of "0", "1", .. etc would be easier to read but  it seems its useful to have the integers as this also checks ordering (which we already check elsewhere but its still a useful reminder).